src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+            {
+                var parameterIndices = new Dictionary<string, int>();
+                int index = 0;
+                foreach (var param in allParameterNames)
+                {
+                    parameterIndices.Add(param, index);
+                    index++;
+                }
+                if (!parameterIndices.ContainsKey(parameterName))
+                {
+                    // something went very wrong
+                    Debug.Assert(false); // TODO message or remove assert
+                    adjacentNameAttribute = null;
+                    adjacentAttributeIsBefore = false;
+                    return;
+                }
+
+                var walker = new FindAdjacentParameterDocumentationWalker(parameterIndices[parameterName], parameterIndices, elementKind);
+                walker.VisitLeadingTrivia(tokenWithDocumentationComment);
+
+                if (walker._documentedParameterBefore != null)
+                {
+                    adjacentNameAttribute = walker._documentedParameterBefore;
+                    adjacentAttributeIsBefore = true;
+                }
+                else
+                {
+                    adjacentNameAttribute = walker._documentedParameterAfter;
+                    adjacentAttributeIsBefore = false;
+                }
+            }
+
+            public override void VisitXmlNameAttribute(XmlNameAttributeSyntax node)
+            {
+                if (node.GetElementKind() == _elementKind)
+                {
+                    int index;
+                    if (_parameterIndices.TryGetValue(node.Identifier.Identifier.ValueText, out index))
+                    {
+                        if (_documentedParameterBeforeIndex < index & index < _documentedParameterAfterIndex)
+                        {
+                            if (index < _parameterIndex)
+                            {
+                                _documentedParameterBefore = node;
+                                _documentedParameterBeforeIndex = index;
+                            }
+                            if (_parameterIndex < index)
+                            {
+                                _documentedParameterAfter = node;
+                                _documentedParameterAfterIndex = index;
+                            }
+                        }
+                    }
+                    else if (_documentedParameterAfter == null)
+                    {
+                        // in case all the param or typeparam tags that exist are not a parameter
+                        _documentedParameterAfter = node;
+                    }
+                }
+                base.VisitXmlNameAttribute(node);
+            }
+        }
+
+        private class AddParameterXmlDocumentationCodeAction : CodeAction
```

Comment:
It'd be good to have separate code action types for parameters versus type parameters. Our telemetry captures the type of the code action. So, if these are the same type, we won't be able to tell the difference between them.